Entities/UpsideDownJumpThru.cs

```diff
@@ -0,0 +1,326 @@
+﻿using Celeste.Mod.Entities;
+using Microsoft.Xna.Framework;
+using Mono.Cecil;
+using Mono.Cecil.Cil;
+using Monocle;
+using MonoMod.Cil;
+using MonoMod.RuntimeDetour;
+using System;
+using System.Collections.Generic;
+using System.Reflection;
+using System.Linq;
+
+namespace Celeste.Mod.SpringCollab2020.Entities {
+    [CustomEntity("SpringCollab2020/UpsideDownJumpThru")]
+    [Tracked]
+    class UpsideDownJumpThru : JumpThru {
+
+        private static FieldInfo actorMovementCounter = typeof(Actor).GetField("movementCounter", BindingFlags.Instance | BindingFlags.NonPublic);
+        private static FieldInfo playerVarJumpTimer = typeof(Player).GetField("varJumpTimer", BindingFlags.Instance | BindingFlags.NonPublic);
+
+        private static ILHook playerOrigUpdateHook;
+
+        public static void Load() {
+            using (new DetourContext { Before = { "*" } }) { // these don't always call the orig methods, better apply them first.
+                // fix general actor/platform behavior to make them comply with jumpthrus.
+                On.Celeste.Actor.MoveVExact += onActorMoveVExact;
+                On.Celeste.Platform.MoveVExactCollideSolids += onPlatformMoveVExactCollideSolids;
+            }
+
+            // fix player specific behavior allowing them to go through upside-down jumpthrus.
+            On.Celeste.Player.OnCollideV += onPlayerOnCollideV;
+
+            // block player if they try to climb past an upside-down jumpthru.
+            IL.Celeste.Player.ClimbUpdate += patchPlayerClimbUpdate;
+
+            // ignore upside-down jumpthrus in select places.
+            playerOrigUpdateHook = new ILHook(typeof(Player).GetMethod("orig_Update"), filterOutJumpThrusFromCollideChecks);
+            IL.Celeste.Player.DashUpdate += filterOutJumpThrusFromCollideChecks;
+            IL.Celeste.Player.RedDashUpdate += filterOutJumpThrusFromCollideChecks;
+        }
+
+        public static void Unload() {
+            On.Celeste.Actor.MoveVExact -= onActorMoveVExact;
+            On.Celeste.Platform.MoveVExactCollideSolids -= onPlatformMoveVExactCollideSolids;
+
+            On.Celeste.Player.OnCollideV -= onPlayerOnCollideV;
+            IL.Celeste.Player.ClimbUpdate -= patchPlayerClimbUpdate;
+
+            if (playerOrigUpdateHook != null)
+                playerOrigUpdateHook.Dispose();

```

Comment:
This (and the preceeding line) can be simplified to
```suggestion
                playerOrigUpdateHook?.Dispose();
```